sdk/dotnet/Pulumi.Automation/LocalWorkspace.cs

```diff
@@ -341,6 +342,31 @@ public static Task<WorkspaceStack> CreateOrSelectStackAsync(LocalProgramArgs arg
             this._readyTask = Task.WhenAll(readyTasks);
         }
 
+        private async Task InitializeProjectSettingsAsync(ProjectSettings projectSettings,
+                                                          CancellationToken cancellationToken)
+        {
+            // If given project settings, we want to write them out to
+            // the working dir. We do not want to override existing
+            // settings with default settings though.
+
+            var existingSettings = await this.GetProjectSettingsAsync(cancellationToken);
+            if (existingSettings == null)
+            {
+                await this.SaveProjectSettingsAsync(projectSettings, cancellationToken);
+            }
+            else
+            {
+                if (!projectSettings.IsDefault)
+                {
+                    // TODO we are overwriting settings found on disk
+                    // with settings passed in through code. Seems
+                    // like a good place to emit a warning, but not
+                    // clear which logging facility to use.
+                    await this.SaveProjectSettingsAsync(projectSettings, cancellationToken);
```

Comment:
Can you explain this a bit? Why are we overwriting settings found on disk with settings passed in through code? 

Settings already on disk should always take precedence. There are 3 possible scenarios:
1. Create an inline program stack that passes in a `WorkDir` LocalWorkspaceOption - in this case we should check if the workdir has a Pulumi.yaml and read it. If that succeeds, do nothing (i.e do not set projectSettings)
2. Create an inline program stack that passes in a `ProjectSettings` LocalWorkspaceOption - in this case, save the passed in `ProjectSettings`.
3. Create an inline program with neither of the above. In this case, create the default project settings and save those.

Although I guess this begs the question of what if both workdir and project settings are passed in. Is that what you're addressing here? 

In my implementation for nodejs/python/go, the existing `Pulumi.yaml` always takes precedence - but I could see how that might also be confusing for the user if they have passed in `ProjectSettings` and they are not respected.

Reply:
Ok let me try that! To clarify, one ya non-default ProjectSettings. If Default ProjectSettings is passed in, then a lot of code that did not bother customizing the settings, including our tests, will start failing, simply because the framework creates a default ProjectSettings.

Reply:
>Are WorkDir and ProjectSettings mutually exclusive? Hm I guess not we might have a user that wants to set both.

Yeah those two settings are not mutually exclusive. Because a `WorkDir` may not have a pre-existing `Pulumi.yaml`. We only run into trouble if the workdir already has settings.

Reply:
>Ok let me try that! To clarify, only non-default ProjectSettings. If Default ProjectSettings is passed in, then a lot of code that did not bother customizing the settings, including our tests, will start failing, simply because the framework creates a default ProjectSettings.

🤔 We should not create default project settings if there is already a `Pulumi.yaml` in the `WorkDir` or if a user passes in `ProjectSettings`.

Reply:
I think the last version gets to that behavior.

It's a bit awkward because of the comparer changes.

I'm poking around to see if there is anything simpler; perhaps we could remove current uses of ProjectSettings.IsDefault and use `null` value as "default", and only coerce null to default when needed.

Reply:
I tried that and it's not pretty, I prefer what's in here; the ProjectSettings currently carries 'required' arguments like ProjectName and runtime (defaulting to nodejs), we'd have to remove null checks and add ProjectName back to something like LocalWorkspaceOptions, it's a bit confusing. I will leave as is for now unless there's more pushback.

Reply:
How it is now is fine with me. 🙏🏽